### PR TITLE
Flask: implement email notifications

### PR DESCRIPTION
**Pull request by hannahle**
- We used SendGrid Email API to send emails to appropriate users when an analysis is requested.
- The `send_email` function will send from whichever user email we specify in the `from_email` field. The email domains in `from_email` would need to be verified for some email providers (like SickKids) to accept it and show it in the end user's inbox. So far, we've verified the `ccmdev.ca` domain, and the test `from_email` is `test@ccmdev.ca`.
- In the email, there would also be an Unsubscribe option, where users can choose to not receive future notification emails. 
- The code also checks for whether we have exceeded the email limit in the day, and if so, queue the email the next day. Otherwise, it sends the email 1 minute after an analysis has been successfully requested.

**To test the code, you need to:**

1. Create a SendGrid account and an API key and add it to `SENDGRID_API_KEY` in `.env`.
2. Follow these steps in [Single Sender Verification](https://docs.sendgrid.com/ui/sending-email/sender-verification) to verify yourself as a sender. If you are not verified as a sender, you will not be able to send any email.Hence, the recipient (which is also you in this case, because you're testing) won't receive any email. 
_Note:_ Single Verification is only used for fast, local testing and not recommended in production. As mentioned above, we are going to verify the email sender domains, similar to `ccmdev.ca`, in production.
3. Add the email you have verified in step 2 to the `from_email` argument in `email.send_email`
4. Add your personal (non-SickKids) email to the `to` field. The reason why you can't use your SickKids email here is because SickKids email server doesn't trust emails sent via sendgrid as a proxy. This problem is solved once we verify the domain. 
5. Import [this](https://drive.google.com/file/d/1qgtVTJbpqtqpjSJoUOlybzjBemeyXHzp/view?usp=sharing) custom email HTML template to the **[Dynamic Templates](https://mc.sendgrid.com/dynamic-templates)** section in your SendGrid dashboard.
6. Copy-paste the template ID and add it to `.env`.
7. Create an Unsubscribe group in your dashboard [here](https://app.sendgrid.com/suppressions/advanced_suppression_manager). Add your unsubscribe group ID to `.env`.
5. Try requesting an analysis and you should see an email to your personal inbox!

**Note:**

This code can be merged into master with email notification disabled by simplying setting the `SENDGRID_API_KEY` to empty until we've properly confirmed all collaborators that we will be sending them email.

### CODE REVIEW

File: docker-compose.yaml
Comment (kevinlul) [inline docker-compose.yaml:47]: Looks like this merge went the wrong way. The result should be
```suggestion
      MINIO_ACCESS_KEY:
      MINIO_SECRET_KEY:
      MINIO_REGION_NAME: 
      SENDGRID_API_KEY:
      SENDGRID_UNSUBSCRIBE_GROUP_ID:
      SENDGRID_EMAIL_TEMPLATE_ID:
```

File: docker-compose.test.yaml
Comment (kevinlul) [inline docker-compose.test.yaml:32]:
```suggestion
```
Shouldn't be sending emails when running tests

File: react/src/hooks/utils.tsx
Comment (kevinlul) [inline react/src/hooks/utils.tsx:18]: What's going on here in the frontend? I think you might have an old version of react-query installed. Try running `yarn` again to update.

File: flask/app/email.py
Comment (kevinlul) [inline flask/app/email.py:91]: This is only used once. Every other time you're directly calling `strftime`

File: flask/app/email.py
Comment (kevinlul) [inline flask/app/email.py:20]: We use `getenv` elsewhere in the codebase, though they are totally aliases.

File: flask/app/email.py
Comment (kevinlul) [inline flask/app/email.py:25]: Add type hints

File: flask/app/analyses.py
Comment (kevinlul) [inline flask/app/analyses.py:457]: It doesn't look like this is conditioned on having a SendGrid API key?

**Review by kevinlul — CHANGES_REQUESTED** [review]
So I'm not quite seeing how email batching is implemented here, just a delay. I also don't see how the metadata dictionary becomes the email body. Are these both done on SendGrid?

File: react/src/hooks/utils.tsx
Comment (hannahle) [reply react/src/hooks/utils.tsx:18]: Cool! Will try installing yarn again.

File: flask/app/analyses.py
Comment (hannahle) [reply flask/app/analyses.py:457]: When the `SENDGRID_API_KEY` in `.env` is set to empty, the `send_email` would immediately throw an 403 Unauthorized error and the email would fail to be sent.

File: flask/app/email.py
Comment (hannahle) [reply flask/app/email.py:20]: > So I'm not quite seeing how email batching is implemented here, just a delay. I also don't see how the metadata dictionary becomes the email body. Are these both done on SendGrid?

Ah, the email batching (i.e. content from 5 analysis requests merged into 1 email) is not implemented yet. One email is sent per analysis request for now. I'll look into this!

Regarding the email body, this is done via SendGrid. What's happening is you have an HTML file that represents the email body and you set that as a template inside SendGrid. To tell your code to use that template, you can see that I have added a template ID.

**Comment by hannahle** [comment]
> So I'm not quite seeing how email batching is implemented here, just a delay. I also don't see how the metadata dictionary becomes the email body. Are these both done on SendGrid?

Ah, the email batching (i.e. content from 5 analysis requests merged into 1 email) is not implemented yet. One email is sent per analysis request for now. I'll look into this!

Regarding the email body, this is done via SendGrid. What's happening is you have an HTML file that represents the email body and you set that as a template inside SendGrid. To tell your code to use that template, you can see that I have added a template ID.

File: flask/app/analyses.py
Comment (kevinlul) [reply flask/app/analyses.py:457]: It's important to have the conditional before that, because you're sending out a meaningless request to another server and that adds to the response time of this endpoint.

Comment (hannahle) [reply flask/app/analyses.py:457]: Cool! I'll add it in.

**Comment by kevinlul** [comment]
#935

File: flask/app/analyses.py
Comment (kevinlul) [inline flask/app/analyses.py:489]:
```suggestion
    current_cache = cache.get("analyses_emails") or []
```
With this you don't need to have the if/else after with nearly identical bodies

File: flask/app/config.py
Comment (kevinlul) [inline flask/app/config.py:41]: Is this actually used?

File: flask/app/__init__.py
Comment (kevinlul) [inline flask/app/__init__.py:48]: We shouldn't be registering this altogether if we don't have an API key. We also definitely need an out-of-process cache as there are multiple Gunicorn workers, and we can't send out a batch of emails on exit because that would happen for every update we deploy and potentially spuriously since the server process could be restarted for whatever reason.

**Review by kevinlul — CHANGES_REQUESTED** [review]
It looks like there are still changes to the `react/` subdirectory. You can revert them with `git checkout origin/master react`

I'm not seeing where the cache is emptied after the email is sent.

File: flask/app/email.py
Comment (kevinlul) [inline flask/app/email.py:25]: Typo, suggest "sends an email based on a template stored in SendGrid"

File: flask/app/email.py
Comment (kevinlul) [inline flask/app/email.py:46]: Don't dump this to logs

File: flask/app/extensions.py
Comment (kevinlul) [inline flask/app/extensions.py:1]: We're no longer using this, so make sure to remove it from dependencies!

File: react/src/functions.tsx
Comment (kevinlul) [inline react/src/functions.tsx:231]: Please make sure there are no changes to the `react/` subdirectory (unless needed for features of this PR). You can revert them with `git checkout origin/master react`

File: flask/app/email.py
Comment (kevinlul) [inline flask/app/email.py:19]: I'm going to have a look at whether setting the timezone in the container causes a problem with our APIs that return timestamps.

**Review by kevinlul — CHANGES_REQUESTED** [review]

File: flask/app/email.py
Comment (kevinlul) [reply flask/app/email.py:19]: It will because our datetime objects are naive, so it's out of scope for this PR. I think this should be set to `America/Toronto` or it might not account for daylight saving?

**Review by kevinlul — APPROVED** [review]
Thanks for working on this big feature. Looks good to me! I'll wait until Monday to merge though since Conor isn't here.

**Pull request merged**
